src/track/beats.cpp

```diff
@@ -1,29 +1,537 @@
-
 #include "track/beats.h"
 
+#include "track/beatutils.h"
+#include "track/track.h"
+
 namespace mixxx {
 
-int Beats::numBeatsInRange(double dStartSample, double dEndSample) {
-    double dLastCountedBeat = 0.0;
+const QString BeatsInternal::BEAT_MAP_VERSION = "BeatMap-1.0";
+const QString BeatsInternal::BEAT_GRID_1_VERSION = "BeatGrid-1.0";
+const QString BeatsInternal::BEAT_GRID_2_VERSION = "BeatGrid-2.0";
+const QString BeatsInternal::BEATS_VERSION = "Beats-1.0";
+
+namespace {
+inline bool TimeSignatureMarkerEarlier(
+        const track::io::TimeSignatureMarker& marker1,
+        const track::io::TimeSignatureMarker& marker2) {
+    return marker1.downbeat_index() < marker2.downbeat_index();
+}
+inline bool BpmMarkerEarlier(
+        const track::io::BpmMarker& marker1,
+        const track::io::BpmMarker& marker2) {
+    return marker1.beat_index() < marker2.beat_index();
+}
+constexpr int kSecondsPerMinute = 60;
+constexpr double kBeatVicinityFactor = 0.1;
+
+inline FrameDiff_t getBeatLengthFrames(Bpm bpm,
+        double sampleRate,
+        TimeSignature timeSignature = TimeSignature()) {
+    return kSecondsPerMinute * sampleRate *
+            (4.0 / timeSignature.getNoteValue()) / bpm.getValue();
+}
+} // namespace
+
+Beats::Beats(const Track* track,
+        const QVector<FramePos>& beats,
+        const QVector<track::io::TimeSignatureMarker>& timeSignatureMarkers)
+        : Beats(track) {
+    if (beats.size() > 0) {
+        m_beatsInternal.initWithAnalyzer(beats, timeSignatureMarkers);
+    }
+    slotTrackBeatsUpdated();
+}
+
+Beats::Beats(const Track* track, const QByteArray& byteArray)
+        : Beats(track) {
+    m_beatsInternal.initWithProtobuf(byteArray);
+    slotTrackBeatsUpdated();
+}
+
+Beats::Beats(const Track* track)
+        : m_mutex(QMutex::Recursive), m_track(track) {
+    // BeatMap should live in the same thread as the track it is associated
+    // with.
+    slotTrackBeatsUpdated();
+    connect(m_track, &Track::beatsUpdated, this, &Beats::slotTrackBeatsUpdated);
+    connect(m_track, &Track::changed, this, &Beats::slotTrackBeatsUpdated);
```

Comment:
I'm quite confused by the signaling between Track and Beats. Beats should signal to Track when Beats is updated, but Track doesn't directly edit Beats so why are signals from Track connected to Beats::slotTrackBeatsUpdated? I'm unclear why slotTrackBeatsUpdated exists. @uklotzde any thoughts on this?

Reply:
Those cyclic signal/slot connections seem to be wrong. The connections to Track::beatsUpdated and Track::changed must be removed.

The Beats object must not react to changes in the Track object. It is the responsibility of the Track object to update or replace the Beats object as needed.

Reply:
@Be-ing Good catch!

Reply:
I think this may mostly be an issue of confusing naming. All that Beats::slotTrackBeatsUpdated does is get the sample rate and duration from the Track. I think this could be replaced by passing these parameters to the Beats constructor. An instance of Beats should not outlive Track or moved between Track instances, so if I understand correctly, there shouldn't be any scenario where Beats needs to get updated information from Track.

Reply:
The track has a pointer to the Beats object and is supposed to update it directly by calling a method those properties are updated. It doesn't matter if the Beats object gets orphaned. Then it won't be updated by the track anymore.

In the long term, only Track should be a QObject that sends signals. Neither Beats nor individual Cue objects are needed.